### PR TITLE
Call a contract

### PR DESCRIPTION
**Pull request by ascjones**
Make the contract call extrinsic work, and update the contracts tests so they can run in parallel and all succeed.

**Note**: the contracts tests require a node-template based runtime with contracts enabled running.

### CODE REVIEW

**Comment by dvdplm** [comment]
This looks interesting. I get `Metadata(ModuleNotFound("Contracts"))` when I try to run these tests, likely because of how I run the node (`client/run.sh`).

**Comment by ascjones** [comment]
Yeah the contracts module is not included in the test runtime at the moment. We could consider adding it, at the cost of increasing complexity of the test node. 

But it does beg the question as to whether we should add *all* modules to that, in which case it would just become `node-runtime`. I don't know the answer to that but we need to figure it out

**Comment by dvdplm** [comment]
> But it does beg the question as to whether we should add _all_ modules to that, in which case it would just become `node-runtime`. I don't know the answer to that but we need to figure it out

Agreed. I'm fine with a heavy-handed interim solution. Can you spell out the pros/cons to using the whole `node-runtime`?

**Comment by ascjones** [comment]
Well now that there exists tagged releases of `substrate` on docker: https://hub.docker.com/r/parity/substrate/tags we could use one of those on CI. 

So pros: don't need a dependency, or to compile the maintain the whole node runtime. Cons: some CI setup.

I think it could work: https://github.com/paritytech/substrate-subxt/issues/169.

**Review by dvdplm — APPROVED** [review]
Good stuff. I think the account generation and nonce handling stuff is useful enough to warrant making it available to other code as well; can be done in a diff PR though.

File: src/frame/contracts.rs [hunk 1]
```diff
@@ -115,44 +116,164 @@ pub struct InstantiatedEvent<T: Contracts> {
...
 }
 
+/// Contract execution event.
+///
+/// Raised upon successful executionFailing  of a contract call
+#[derive(Clone, Debug, Eq, PartialEq, Event, Decode)]
+pub struct ContractExecutionEvent<T: Contracts> {
+    /// Caller of the contract.
+    pub caller: <T as System>::AccountId,
+    /// Raw contract event data
```
[the remark this answers is not in the document]
Comment (ascjones) [reply src/frame/contracts.rs:126]: The data is the encoded contract event, I will update the comment

File: src/frame/contracts.rs [hunk 2]
```diff
@@ -115,44 +116,164 @@ pub struct InstantiatedEvent<T: Contracts> {
...
+
+            let result = self.client.put_code_and_watch(&self.signer, &code).await?;
+            let code_stored = result.code_stored()?.ok_or_else(|| {
+                Error::Other("Failed to find a CodeStored event".into())
+            })?;
+            log::info!("Code hash: {:?}", code_stored.code_hash);
+            Ok(code_stored)
+        }
+
+        async fn instantiate(
```
[the remark this answers is not in the document]
Comment (ascjones) [reply src/frame/contracts.rs:224]: The function names should match with the contract pallet extrinsics

**Comment by ascjones** [comment]
> the account generation and nonce handling stuff is useful enough to warrant making it available to other code as well; can be done in a diff PR though.

https://github.com/paritytech/substrate-subxt/issues/176

**Review by dvdplm — APPROVED** [review]

**Pull request merged**